File: backend/src/main/java/info/gamewise/lor/videos/entity/VideoJpaEntity.java
package info.gamewise.lor.videos.entity;

import com.github.brunosc.fetcher.domain.VideoThumbnails;
import com.github.brunosc.lor.domain.LoRRegion;
import info.gamewise.lor.videos.domain.json.Champion;
import info.gamewise.lor.videos.port.out.VideosNotInDatabasePort.NewVideo;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.Set;

import static java.util.stream.Collectors.toUnmodifiableSet;

@Document(collection = "videos")
class VideoJpaEntity {

    @Id
    private String id;

    @Indexed(unique = true)
    private String videoId;

    private String url;
    private String title;
    private String deckCode;
    private String channel;
    private Set<LoRRegion> regions;
    private Set<String> champions;
    private LocalDateTime publishedAt;
    private VideoThumbnails thumbnails;

    VideoJpaEntity() {

    }

    VideoJpaEntity(NewVideo newVideo) {
        this.videoId = newVideo.details().getId();
        this.url = newVideo.details().getUrl();
        this.title = newVideo.details().getTitle();
        this.deckCode = newVideo.deckCode();
        this.publishedAt = newVideo.details().getPublishedAt();
        this.thumbnails = newVideo.details().getThumbnails();
        this.channel = newVideo.channel().code();
        this.regions = newVideo.regions();
        this.champions = newVideo.champions().stream().map(Champion::code).collect(toUnmodifiableSet());
    }

    public String getId() {
        return id;
    }

    public String getVideoId() {
        return videoId;
    }

    public String getUrl() {
        return url;
    }

    public String getTitle() {
        return title;
    }

    public String getChannel() {
        return channel;
    }

    public Set<LoRRegion> getRegions() {
        return regions;
    }

    public Set<String> getChampions() {
        return champions;
    }

    public LocalDateTime getPublishedAt() {
        return publishedAt;
    }

    public VideoThumbnails getThumbnails() {
        return thumbnails;
    }

    public String getDeckCode() {
        return deckCode;
    }
}

File: backend/src/main/kotlin/info/gamewise/lor/videos/entity/video/VideoJpaEntity.kt
package info.gamewise.lor.videos.entity.video

import com.github.brunosc.fetcher.domain.VideoThumbnails
import com.github.brunosc.lor.domain.LoRRegion
import info.gamewise.lor.videos.port.out.NewVideo
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDateTime

@Document(collection = "videos")
internal class VideoJpaEntity(@Id var id: String? = null,
                              @Indexed(unique = true) var videoId: String? = null,
                              var url: String? = null,
                              var title: String? = null,
                              var deckCode: String? = null,
                              var channel: String? = null,
                              var regions: Set<LoRRegion>? = null,
                              var champions: Set<String>? = null,
                              var publishedAt: LocalDateTime? = null,
                              var thumbnails: VideoThumbnails? = null) {


    constructor(newVideo: NewVideo) : this(
        videoId = newVideo.details.id,
        url = newVideo.details.url,
        title = newVideo.details.title,
        deckCode = newVideo.deckCode,
        publishedAt = newVideo.details.publishedAt,
        thumbnails = newVideo.details.thumbnails,
        channel = newVideo.channel.code,
        regions = newVideo.regions,
        champions = newVideo.champions.map { it.code }.toSet()
    )
}